crates/etc-merge/src/lib.rs

```diff
@@ -0,0 +1,1079 @@
+                }
+
+                Err(e) => Err(e).context(format!("{path:?}"))?,
+            },
+        }
+
+        current_path.pop();
+    }
+
+    Ok(())
+}
+
+/// Traverses and collects directory trees for three etc states.
+///
+/// Recursively walks through the given *pristine*, *current*, and *new* etc directories,
+/// building filesystem trees that capture files, directories, and symlinks.
+/// Device files, sockets, pipes etc are ignored
+///
+/// It is primarily used to prepare inputs for later diff computations and
+/// comparisons between different etc states.
+///
+/// # Arguments
+///
+/// * `pristine_etc` - The reference directory representing the unmodified version or current /etc.
+/// Usually this will be obtained by remounting the EROFS image to a temporary location
+///
+/// * `current_etc` - The current `/etc` directory
+///
+/// * `new_etc` - The directory representing the `/etc` directory for a new deployment. This will
+/// again be usually obtained by mounting the new EROFS image to a temporary location. If merging
+/// it will be necessary to make the `/etc` for the deployment writeable
+///
+/// # Returns
+///
+/// [`anyhow::Result`] containing a tuple of directory trees in the order:
+///
+/// 1. `pristine_etc_files` – Dirtree of the pristine etc state
+/// 2. `current_etc_files` – Dirtree of the current etc state
+/// 3. `new_etc_files` – Dirtree of the new etc state
+pub fn traverse_etc(
+    pristine_etc: &CapStdDir,
+    current_etc: &CapStdDir,
+    new_etc: &CapStdDir,
+) -> anyhow::Result<(
+    Directory<CustomMetadata>,
+    Directory<CustomMetadata>,
+    Directory<CustomMetadata>,
+)> {
+    let mut pristine_etc_files = Directory::default();
+    recurse_dir(pristine_etc, &mut pristine_etc_files)
+        .context(format!("Recursing {pristine_etc:?}"))?;
+
+    let mut current_etc_files = Directory::default();
+    recurse_dir(current_etc, &mut current_etc_files)
+        .context(format!("Recursing {current_etc:?}"))?;
+
+    let mut new_etc_files = Directory::default();
+    recurse_dir(new_etc, &mut new_etc_files).context(format!("Recursing {new_etc:?}"))?;
+
+    return Ok((pristine_etc_files, current_etc_files, new_etc_files));
+}
+
+/// Computes the differences between two directory snapshots.
+#[context("Computing diff")]
+pub fn compute_diff(
+    pristine_etc_files: &Directory<CustomMetadata>,
+    current_etc_files: &Directory<CustomMetadata>,
+) -> anyhow::Result<Diff> {
+    let mut diff = Diff {
+        added: vec![],
+        modified: vec![],
+        removed: vec![],
+    };
+
+    get_modifications(
+        &pristine_etc_files,
+        &current_etc_files,
+        PathBuf::new(),
+        &mut diff,
+    )?;
+
+    get_deletions(
+        &pristine_etc_files,
+        &current_etc_files,
+        PathBuf::new(),
+        &mut diff,
+    )?;
+
+    Ok(diff)
+}
+
+/// Prints a colorized summary of differences to standard output.
+pub fn print_diff(diff: &Diff) {
```

Comment:
Seems a bit more elegant to have the caller pass the stream to write to; this allows easy unit testing among other things